Write marshalled kubeletconfig object to init-config-dir

xiangpengzhao opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
from @luxas :
>Write the the marshalled kubeletconfig object to /var/lib/kubelet/config/init/kubelet so that the kubelet will start up with the right params on init/join. The only params expected in the kubelet command-line after this is kubelet --init-config-dir /var/lib/kubelet/config/init --dynamic-config-dir /var/lib/kubelet/config/dynamic

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: https://github.com/kubernetes/kubeadm/issues/28#issuecomment-345502933

**Special notes for your reviewer**:
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

xiangpengzhao committed:
Write marshalled kubeletconfig object to init-config-dir

xiangpengzhao committed:
Auto generated BUILD files.

luxas commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -155,6 +155,12 @@ const (
+	KubeletBaseConfigurationDir = "/var/lib/kubelet/config/init"
+
+	// KubeletBaseConfigurationFile specifies the file name on the node which stores initial remote configuration of kubelet
+	KubeletBaseConfigurationFile = "kubelet.yaml"
```
This constant is defined somewhere in the kubelet codebase. Right now it is `kubelet` I think. You can check the PR where @mtaufen is in process to graduate file-loading from alpha to beta

xiangpengzhao replied:
OK, I will check it. Thanks for pointing it out!

xiangpengzhao replied:
Updated. I add a TODO here.

https://github.com/kubernetes/kubernetes/pull/53833/files#diff-fbd96764b3922031c77bb5efdab3269aR21

luxas commented on cmd/kubeadm/app/phases/kubelet/kubelet.go:
```diff
@@ -86,6 +90,15 @@ func UpdateNodeWithConfigMap(client clientset.Interface, nodeName string) error
 			return false, nil
 		}
 
+		baseCongfig, err := yaml.Marshal(kubeletCfg.Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey])
```
I think you misunderstood this. The configuration that should be written to the file is `cfg.KubeletConfiguration.BaseConfig`; basically write the same thing you have in the configmap on disk in the init code.

In the join code, this config should be looked up from the configmap (we need a new RBAC rule for that)

xiangpengzhao replied:
The `kubeletCfg` here is the configmap we stored in API server, it's `Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey]` is the BaseConfig which we want.

See Line 50:
```go
kubeletBytes, err := kubeadmutil.MarshalToYamlForCodecs(cfg.KubeletConfiguration.BaseConfig, kubeletconfigv1alpha1.SchemeGroupVersion, *kubeletCodecs)
```
and Line 60-61:
```go
 		Data: map[string]string{
 			kubeadmconstants.KubeletBaseConfigurationConfigMapKey: string(kubeletBytes),
 		},
```

Feel free to correct me if I'm indeed misunderstood here :)

xiangpengzhao replied:
>we need a new RBAC rule for that

What RBAC rule is needed? The rules for group "system:nodes" have already been created in init (`CreateBaseKubeletConfiguration`). Am I missing sth here?
```go
 	if err := createKubeletBaseConfigMapRBACRules(client); err != nil {
 		return fmt.Errorf("error creating base kubelet configmap RBAC rules: %v", err)
 	}
```

mtaufen replied:
wrt RBAC, see the instructions regarding permissions here: https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/

xiangpengzhao replied:
Does that mean we should create separate rules for each node?

luxas replied:
Ok, I'll clarify this and alter my earlier plan a bit to make it better.
Let me know if this sounds good to you

1) There should be two functions: one that takes a `string` to write to disk (`WriteInitKubeletConfigToDisk`) and one for `kubeadm join` that takes a k8s client which extracts the config string (`GetKubeletConfigFromConfigMap`)
2) `WriteInitKubeletConfigToDisk` should:
 - Write the string down to this file `filepath.Join(KubeletBaseConfigurationDir, KubeletBaseConfigurationFile)`
3) `GetKubeletConfigFromConfigMap` should:
 - Get the base-config ConfigMap from the API server using a client that is passed in
 - Extract the value string from the `kubeadmconstants.KubeletBaseConfigurationConfigMapKey`

No `yaml.Marshal` or similar code should be used here.

xiangpengzhao commented:
bazel test failed. Seems like there is no write privilege to `/var/lib/kubelet/` ? @luxas any ideas?

```
//cmd/kubeadm/app/phases/kubelet:go_default_test 0.00s
bazel test //cmd/kubeadm/app/phases/kubelet:go_default_test
--- FAIL: TestCreateBaseKubeletConfiguration (0.52s)
	kubelet_test.go:83: CreateBaseKubeletConfiguration: unexepected error failed to write initial remote configuration of kubelet into file "/var/lib/kubelet/config/init/kubelet.yaml" for node fake-node: open /var/lib/kubelet/config/init/kubelet.yaml: no such file or directory
--- FAIL: TestUpdateNodeWithConfigMap (0.50s)
	kubelet_test.go:118: UpdateNodeWithConfigMap: unexepected error failed to write initial remote configuration of kubelet into file "/var/lib/kubelet/config/init/kubelet.yaml" for node fake-node: open /var/lib/kubelet/config/init/kubelet.yaml: no such file or directory
FAIL
```

luxas reviewed (commented):
@xiangpengzhao The overall flow looks like this for `kubeadm join`:

```
1) token discovery as usual => `/etc/kubernetes/bootstrap-kubelet.conf` that will be used for TLS bootstrapping
2) get the base kubelet config ConfigMap using the bootstrap token as the credential
3) write the init config down to disk so the kubelet can start
4) The kubelet starts and performs the TLS bootstrap
5) kubeadm PATCHes the node to start using DKC using the kubelet's TLS bootstrapped credential
```

In `kubeadm init` this will happen:
```
... normal tasks like certs, kubeconfig, static pods, etc
1, NEW) Write the marshalled kubeletconfig down to the init config location
... now the kubelet can start the control plane
... control plane is up 
2) Create the kubeletconfig ConfigMap
3, EDIT) Allow `system:nodes` AND `system:bootstrappers:kubeadm:default-node-token` to access the ConfigMap. This is so `kubeadm join` can get the init config in step 2) above.
4) PATCH the master node to use Dynamic Kubelet Config
```

xiangpengzhao committed:
Refactoring writing kubeletconfig.

xiangpengzhao commented:
@luxas thanks for clarification! updated. Hope this is close to what you expected :)

luxas commented on cmd/kubeadm/app/phases/kubelet/kubelet.go:
```diff
@@ -51,6 +56,10 @@ func CreateBaseKubeletConfiguration(cfg *kubeadmapi.MasterConfiguration, client
 		return err
 	}
 
+	if err := writeInitKubeletConfigToDisk(kubeletBytes); err != nil {
```
This won't work. You need to write down this config RIGHT AFTER the controlplane/etcd phases in order for the kubelet to even start. So please extract this logic into a function you can call in `init.go` after the controlplane stuff, but before you start waiting for the API to be available

luxas commented on cmd/kubeadm/app/phases/kubelet/kubelet.go:
```diff
@@ -152,5 +180,51 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
+		Subjects: []rbac.Subject{
+			{
+				Kind: rbac.GroupKind,
+				Name: kubeadmconstants.NodeBootstrapTokenAuthGroup,
```
Thanks :+1:

luxas reviewed (commented):
Thanks! The only thing that needs fixing yet is that writing down the init config has to happen before waiting on the kubelet to launch the Static Pods, otherwise this will deadlock. But after you've re-organized this to write down the config after the controlplane phase, this should work well :)

One or two unit tests here wouldn't hurt though ;)

luxas commented:
Please fix unit failure: `//cmd/kubeadm/app/phases/kubelet:go_default_test 0.00s`

luxas commented on cmd/kubeadm/app/phases/kubelet/kubelet.go:
```diff
@@ -152,5 +180,51 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
+		return err
+	}
+
+	return apiclient.CreateOrUpdateClusterRoleBinding(client, &rbac.ClusterRoleBinding{
```
Everything here can be just switched for one more subject in the above rolebinding -- just add the subject here to the RoleBinding above and it should work. That should grant `kubeadm join` using the token the right to get the configmap, right?

xiangpengzhao commented:
comments addressed. unit test not updated yet :)

xiangpengzhao committed:
Fix deadlock of writing file on node.

xiangpengzhao commented:
/test pull-kubernetes-e2e-gce

@luxas I have to be off my keyboard for hours. Will update later.Can we catch up the 1.9 train?

luxas commented on cmd/kubeadm/app/phases/kubelet/kubelet.go:
```diff
@@ -148,9 +171,61 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
+}
+
+func writeInitKubeletConfigToDisk(kubeletConfig []byte) error {
+	baseCongfigFile := filepath.Join(kubeadmconstants.KubeletBaseConfigurationDir, kubeadmconstants.KubeletBaseConfigurationFile)
```
nit: `baseConfigFile`

luxas commented on cmd/kubeadm/app/phases/kubelet/kubelet.go:
```diff
@@ -148,9 +171,61 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
+
+func writeInitKubeletConfigToDisk(kubeletConfig []byte) error {
+	baseCongfigFile := filepath.Join(kubeadmconstants.KubeletBaseConfigurationDir, kubeadmconstants.KubeletBaseConfigurationFile)
+	if err := ioutil.WriteFile(baseCongfigFile, kubeletConfig, 0644); err != nil {
```
you should create the directory here with `os.MkdirAll()` before trying to write the file

luxas reviewed (approved):
/lgtm
Two small-ish nits, but I verified that this works well locally as well :+1:

xiangpengzhao committed:
Should make dir before writing file.

xiangpengzhao commented:
Addressed comments, especially https://github.com/kubernetes/kubernetes/pull/56020#discussion_r152578255 in case it cause issues. @luxas PTAL. Thanks!

luxas reviewed (approved):
/lgtm
